src/scalar/golden_ratio.rs

```diff
@@ -0,0 +1,145 @@
+//! Golden ratio is to minimization what bisection is to root finding.
+//! GoldenRatio searches within an interval for a local minimum. At 
+//! every iteration, the interval is decreased in size by a constant 
+//! factor, until the desired precision is obtained.
+//! 
+//! This algorithm is guaranteed to converge on a local minimum in a 
+//! finite amount of steps under very light smoothess criteria for the
+//! target function.
+//! 
+//! In an iteration, the target function is calculated at 4 points:
+//!         +---------+----+---------+
+//! iter 1  a         b    c         d
+//! The interval for the next iteration is chosen to be [a,c] if b<c,
+//! and [b,d] if c<=b. The distances b-a, c-b, and d-c are chosen in such
+//! a way that 3 out of 4 points can be reused, and only 1 new function 
+//! evaluation is required in the next iteration. If b<c this looks like: 
+//!         +---------+----+---------+
+//! iter 1  a         b    c         d
+//!         +----+----+----+
+//! iter 2  a    b    c    d
+
+#[derive(Builder, Debug)]
+pub struct GoldenRatio {
```

Comment:
Infinite search is not possible because the bracket boundary search stops at infinity, which takes around 1027 iterations with `f64` (exact #iters depends on x0 but does not exceed 2023). I agree that we should give more control over the number of iterations/fevals in the explore step, but when memoization lands this depends once more directly on the total number of iterations. 
I feel it is better not to expose an API if we know it will be deprecated soon.